dimagi/commcare-hq
# Question Tile MVP

## Product Description

Adds ability to layout questions horizontally by assigning a 'n-per-row' appearance attribute to questions where "n" is any integer (i.e. "3-per-row"). The question tile width will reduce accordingly and sequential tiles will be displayed such that the sum of the widths of each question tile in the row will be <=12 (12 is from the grid defined in Bootstrap)

https://github.com/dimagi/commcare-hq/assets/88759246/4ce26f0b-1a2f-4908-99b1-1bb5232816e5

![Screen Shot 2023-08-08 at 11 20 16 AM](https://github.com/dimagi/commcare-hq/assets/88759246/53071a78-112d-4b68-9ab0-c2c3b22b51a0)


## Technical Summary

Jira Ticket: https://dimagi-dev.atlassian.net/browse/USH-3418
Design Doc: https://docs.google.com/document/d/1WujrvTByNZaWvauR4e3Md-LLxt0qUpd451iB0PyHoo4/edit#heading=h.ipgou4bu0qoj

For defining the widths of the question tile, we utilize Bootstraps built in [grid](https://getbootstrap.com/docs/5.0/layout/grid/) which defines 12 columns. 

To display sequential question tiles with a combined "12" width on the same row and have following question tiles that do not fit overflow into the next row, each "12" width group is nested in a `<div class="row">`. The class `row` is leveraged from bootstrap. 
On the view model side, I added a `GroupedQuestionTileRow` where each instance contains questions that would all belong in a row. How the form_ui is is built is that a form, question, group, or repeat-juncture objects are defined in a json where objects would be nested in a `children`. I.e. the form contains a group child and the group contains questions children. To create the `GroupedQuestionTileRow` I recursively iterated through the json and grouped questions contained in `children` accordingly (The `GroupedQuestionTileRow` could contain only one question if the question is intended to take up the entire "12" width). As a result, this `GroupedQuestionTileRow` layer is added so a question will always be nested within a GroupedQuestionTileRow.

## Feature Flag

None
## Safety Assurance

### Safety story

Local testing. This change does impact all web app forms.

### Automated test coverage


Tests already exist that entries in form are in the order as expected for the "Form" object. 
Added tests to check questions are grouped correctly when nested within the children of a "Form" Object

### QA Plan


Will pass through QA to check layout of forms with various question types and interactions.
https://dimagi-dev.atlassian.net/browse/QA-5472

### Rollback instructions




### Labels & Review
